### ISSUE TITLE
Impact of scaling on feature attributions

### ISSUE THREAD

**Issue by k-priyadarshi** [kundajelab/deeplift#99]
Hi there,

I am using DeepLIFT on a regression problem where I am predicting continuous values from some continuous features. When I use MinMax scaling on features as well as output, the individual attributions (after bringing them to normal scale) don't sum to the output (on normal scale). So, should DeepLIFT be used without scaling any of the fetaures or output? Because in that case, the feature contributions sum up to the output value. Any suggestions?

**Comment by AvantiShri** [comment]
Hi @k-priyadarshi, so with deeplift the contributions will sum up to the *difference* in the output from its baseline value. If the baseline value of the output is nonzero, then the contributions won’t sum up to the output. Does that make sense?

**Comment by k-priyadarshi** [comment]
> Hi @k-priyadarshi, so with deeplift the contributions will sum up to the _difference_ in the output from its baseline value. If the baseline value of the output is nonzero, then the contributions won’t sum up to the output. Does that make sense?
> 
> (The baseline value of the output is defined as the value of the output when the network is supplied the baseline input)

Hi,

Thanks so much for the prompt reply. Yes, Sum of the feature contributions and baseline output should sum to the actual output. But in the case when MinMaxScaler is used, they don't. They do sum to actual output when no scaling is used for features and output. Basically this has to do with the denominator term in MinMaxScaler which scales the attributions disproprotionally and thus, they don't sum to output value. So, I guess DeepLIFT for regressions should be run without scaling. It would be useful if you recommended something.

Thanks :)

**Comment by AvantiShri** [comment]
Hmm, I think I might understand the problem better with some example code. Are you using https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html?

You mentioned that MinMaxScaler “scales the attributions disproportionately” - this line is a bit confusing to me because I am used to MinMaxScaler being applied as a preprocessing step to the inputs - but here it sounds like MinMaxScaler is being applied to the attributions? Does that mean you are rescaling the attributions with MinMaxScaler after you compute the attributions? Or am I misinterpreting?

**Comment by k-priyadarshi** [comment]
Hi Avanti,

Yes, I am rescaling the attributions to the original scale using MinMaxScaler. Let me explain it with an example.

Suppose I have this one example input:
x = [23, 45, 2, 1000] # 4 features
y = 28 # output

Training Regression after MinMaxScaling x and y:
x_scaled = [0.67, 0.65, 0.04, 0.34]
y_scaled = [0.77]

Model Output, y_hat = 0.765

DeepLIFT Attributions: [0.2, 0.3, 0.1, 0.1] 
baseline_output = 0.065

Now Applying MinMaxScaler to y_hat --> y_hat_original_scale = 27.5 # predicted_output
Now Applying MinMaxScaler to Attributions and baseline_output [according to Scaler for y_hat] doesn't sum up to 27.5.

I don't get this problem when training with unscaled data.

Can you help in suggesting the right method for this?

Thanks!

**Comment by AvantiShri** [comment]
I think the appropriate transformation is the following:

```
new_baseline = old_baseline*(y_max-y_min) + y_min
new_attributions = old_attributions * (y_max-y_min)
```

From your example, it looks like `y_min` is -49 and `y_max-y_min` is 100, i.e. `(28+49)/100 = 0.77` and `(27.5+49)/100 = 0.765`. The transformation would therefore be:
`new_baseline = 0.065*100 - 49 = -42.5`
`new_attributions = [0.2, 0.3, 0.1, 0.1]*100 = [20, 30, 10, 10]`

Now if we do new_baseline + sum(new_attribution), we get -42.5 + 70 = 27.5, which is y_hat_original_scale.

Basically, you should scale the attributions and the baseline by (y_max-y_min), but then only add y_min to the baseline. This is because, if you think about the way the MinMaxScaling is done, y_min is like a baseline value that is added. Does that make sense?

**Comment by k-priyadarshi** [comment]
Hi Avanti,

Thanks a lot for the clarification. I get it now :)
